src/training/train.cpp

```diff
@@ -157,27 +171,85 @@ void TestInference(RecordInput* input, GraphInference* inference) {
-  inference->SSVMInit(FLAGS_regularization_const, FLAGS_svm_margin);
+void Train(RecordInput* input, GraphInference* inference, int fold_id) {
+  inference->CommonInit(FLAGS_regularization_const);
+  if (FLAGS_train_pseudolikelihood == true) {
+    inference->PLInit(FLAGS_beam_size);
+  } else if (FLAGS_train_pseudolikelihood_ssvm == true) {
+    inference->PLInit(FLAGS_beam_size);
+    inference->SSVMInit(FLAGS_svm_margin);
+  } else {
+    inference->SSVMInit(FLAGS_svm_margin);
+  }
   double learning_rate = FLAGS_start_learning_rate;
-  LOG(INFO) << "Starting training with --start_learning_rate=" << std::fixed << FLAGS_start_learning_rate
-      << ", --regularization_const=" << std::fixed << FLAGS_regularization_const
-      << " and --svm_margin=" << std::fixed << FLAGS_svm_margin;
+  if (FLAGS_train_pseudolikelihood == true) {
+    LOG(INFO) << "Starting training using pseudolikelihood as objective function with --start_learning_rate=" << std::fixed << FLAGS_start_learning_rate
+        << ", --regularization_const=" << std::fixed << FLAGS_regularization_const
+        << " and --beam_size=" << std::fixed << FLAGS_beam_size;
+  } else if (FLAGS_train_pseudolikelihood_ssvm == true){
+    LOG(INFO) << "Starting training using first pseudolikelihood as objective function and then change to SSVM after " << std::fixed << FLAGS_num_pass_change_training
+            << "with --start_learning_rate=" << std::fixed << FLAGS_start_learning_rate
+            << ", --initial_learning_rate_ssvm=" << std::fixed << FLAGS_initial_learning_rate_ssvm
+            << ", --regularization_const=" << std::fixed << FLAGS_regularization_const
+            << " and --beam_size=" << std::fixed << FLAGS_beam_size;
+  } else {
+    LOG(INFO) << "Starting SSVM training with --start_learning_rate=" << std::fixed << FLAGS_start_learning_rate
+         << ", --regularization_const=" << std::fixed << FLAGS_regularization_const
+         << " and --svm_margin=" << std::fixed << FLAGS_svm_margin;
+  }
+
   double last_error_rate = 1.0;
+  double initial_learning_rate = learning_rate;
   for (int pass = 0; pass < FLAGS_num_training_passes; ++pass) {
     double error_rate = 0.0;
 
     GraphInference backup_inference(*inference);
 
     int64 start_time = GetCurrentTimeMicros();
     PrecisionStats stats;
-    ParallelForeachInput(input, [&inference,&stats,learning_rate](const Json::Value& query, const Json::Value& assign) {
+    if (FLAGS_train_pseudolikelihood == true ||
+        (FLAGS_train_pseudolikelihood_ssvm == true && pass < FLAGS_num_pass_change_training)) {
+      if (FLAGS_learning_prop_sqrt_pass == true) {
+        learning_rate /= pow(pass + 1, 0.5);
+      } else if (FLAGS_learning_prop_pass == true) {
+        learning_rate /= (pass + 1);
+      } else if (FLAGS_learning_prop_pass_and_initial_learn == true) {
+        learning_rate = initial_learning_rate / (1 + FLAGS_pl_lambda * (pass + 1));
+      }
+    }
+
+    if (FLAGS_profile) {
```

Comment:
I guess you can remove the profiling. Not everyone will download gprofiler in order to compile Nice2Predict. Don't forget to remove the include and also include the fold_id param.